source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+      output = AnalyzeMultiplyOp(inst);
+      break;
+    }
+    default: {
+      output = CreateValueUnknownNode(inst);
+      instruction_map_[inst] = output;
+      break;
+    }
+  }
+  return output;
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeConstant(const ir::Instruction* inst) {
+  if (inst->NumInOperands() != 1) {
+    assert(false);
+  }
+
+  int64_t value = 0;
+
+  // Look up the instruction in the constant manager.
+  const opt::analysis::Constant* constant =
+      context_->get_constant_mgr()->FindDeclaredConstant(inst->result_id());
+
+  if (constant->AsIntConstant()->type()->AsInteger()->IsSigned()) {
+    value = constant->AsIntConstant()->GetS32BitValue();
+  } else {
+    value = constant->AsIntConstant()->GetU32BitValue();
+  }
+
+  return CreateConstant(value);
+}
+
+// Handles both addition and subtraction. If the |sub| flag is set then the
+// addition will be op1+(-op2) otherwise op1+op2.
+SENode* ScalarEvolutionAnalysis::AnalyzeAddOp(const ir::Instruction* add,
+                                              bool sub) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+  std::unique_ptr<SENode> add_node{new SEAddNode(this)};
+
+  add_node->AddChild(
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(0))));
+
+  SENode* op2 =
+      AnalyzeInstruction(def_use->GetDef(add->GetSingleWordInOperand(1)));
+
+  // To handle subtraction we wrap the second operand in a unary negation node.
+  if (sub) {
+    op2 = CreateNegation(op2);
+  }
+  add_node->AddChild(op2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzePhiInstruction(
+    const ir::Instruction* phi) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
+
+  // Get the basic block this instruction belongs to.
+  ir::BasicBlock* basic_block =
+      context_->get_instr_block(const_cast<ir::Instruction*>(phi));
+
+  // And then the function that the basic blocks belongs to.
+  ir::Function* function = basic_block->GetParent();
+
+  // Use the function to get the loop descriptor.
+  ir::LoopDescriptor* loop_descriptor = context_->GetLoopDescriptor(function);
+
+  // We only handle phis in loops at the moment.
+  if (!loop_descriptor) return CreateCantComputeNode();
+
+  // Get the innermost loop which this block belongs to.
+  ir::Loop* loop = (*loop_descriptor)[basic_block->id()];
+  if (!loop) return CreateCantComputeNode();
+
+  std::unique_ptr<SERecurrentNode> phi_node{new SERecurrentNode(this, loop)};
+  instruction_map_[phi] = phi_node.get();
+
+  for (uint32_t i = 0; i < phi->NumInOperands(); i += 2) {
+    uint32_t value_id = phi->GetSingleWordInOperand(i);
+    uint32_t incoming_label_id = phi->GetSingleWordInOperand(i + 1);
+
+    ir::Instruction* value_inst = def_use->GetDef(value_id);
+    SENode* value_node = AnalyzeInstruction(value_inst);
+
+    // Both operands must be loop invariant.
+    if (!IsLoopInvariant(loop, value_node)) {
```

Comment:
I don't understand how the value coming from the latch block can be loop invariant for the simple case:

```
for( int i = 0; i < n; i++ ) {}
```
In this case the result of the increment (which is the value in the OpPhi) is not loop invariant.  Is this a case we are not handling?

Reply:
Well, there will be one invariant operand in this case.  But, I agree, it's confusing what's being looked for here.  It seems like we'll always return a CantCompute node.